### ISSUE TITLE
Delete gtk 2.0 code?

### ISSUE THREAD

**Issue by yetist** [crosswire/xiphos#963]
8 years have passed since the first release of gtk 3.0 in 2011. I think it is time to drop support for gtk2.0. This way we can clean up the code and help to better maintain the existing code.

**Comment by karlkleinpaste** [comment]
that will depend entirely on the level of support for gtk3 in windows, namely in how well it displays. up to this point, the quality of display of gtk3 under a windows build has been nothing less than atrocious. gtk2 is perhaps ancient, but the mingw builds of it work and display very well.

**Comment by yetist** [comment]
Are you telling the truth? Is there a same display problem with [gimp-2.10](https://download.gimp.org/mirror/pub/gimp/v2.10/windows/gimp-2.10.12-setup-3.exe) and [gtk-3.20](https://download.gnome.org/binaries/win64/gedit/gedit-x86_64-3.20.1.msi) running on windows?

**Comment by karlkleinpaste** [comment]
if you're going to accuse me of lying, this discussion will be short.

i hadn't even bothered trying to build for windows using gtk3 in a couple years, because the running result looked so atrocious up to then, and there seemed to be no change toward improving it. gtk2 visuals are easily controlled using c:/Program Files/CrossWire/Xiphos/etc/gtk-2.0/gtkrc. gtk3 never seemed to offer something similar, that we could find when trying to get win32+gtk3 going.

feel free to give it a try yourself and see how it looks. maybe things have changed in recent mingw gtk3 builds. maybe facilities for controlling visuals in win32 w/gtk3 are up to par again. i don't know because i haven't looked in quite a while.

**Comment by yetist** [comment]
Sorry, I have no intention of accusing you of lying, just saying that what you said may not be true.
In gtk3, similar to gtkrc is gtksettings. The theme and other things can be set via settings.ini, although I don't know what you have, but I think it can be fixed.

**Comment by karlkleinpaste** [comment]
when we can buld again for windows using gtk3, and get a good-looking display, i'll be ok with removing the gtk2 code.

**Comment by domcorbex** [comment]
I'm afraid building for windows using gtk3 is not coming soon, the last comment of issue [Update webkitgtk-3.0 to webkit2gtk-4.0 #3484](https://github.com/msys2/MINGW-packages/issues/3484) on msys2/MINGW-packages says:

> It's unlikely that we'll manage to get this building and maintained without upstream officially supporting Windows.

**Comment by greg-hellings** [comment]
Yeah, it's monumentally frustrating. Especially since WebKit officially supports building on Windows out of the box. How difficult would it be, I wonder, to include the native WebKit widget directly into a GTK app?

**Comment by greg-hellings** [comment]
Alternatively, Chromium Embedded Framework exists and works directly out of the box with Windows. Sorry to say, but WebKit is falling by the wayside in favor of Blink/WebEngine. I'm not sure if anything but Safari and Ephiphany still use it and all of them have dropped Windows support from what I can tell.

But that's part of the upgrade treadmill that @karlkleinpaste hates. Here's a guy who demonstrates wrapping Blink into GTK apps including building for Windows with MinGW: https://github.com/cztomczak/cefcapi

**Comment by yetist** [comment]
webkitgtk-3.0 can be compiled with gtk3, it does not necessarily require gtk2, which is not the real reason why we cannot delete the gtk2 code.

**Comment by yetist** [comment]
The real problem is that no suitable webkit2gtk package is available under windows, while gtkhtml and webkitgtk-3.0 under LInux have been deprecated, only webkit2gtk is available.

**Comment by greg-hellings** [comment]
The problem is: Webkit API v1 is deprecated everywhere and removed in most places. Newer versions of WebKitGTK (anything later than 2.4, current is 2.18) officially don't care about Windows. GtkHTML has been a goner for a long time, as well.

If we want a single display mechanism, we

1) Have to accept that our app will forever be ancient and antiquated on Windows, sporting oodles of active CVEs and worse.
2) We need to find an alternative workaround.

* We could embed Blink directly with the Chromium Embedding Framework, which is compatible with Windows and GTK.
* We could go with official builds of WebKit for Windows, but that would mean we need to give up the Fedora cross-build ecosystem (With Github Actions and Travis CI both offering free build time on Windows machines, this isn't an impossible task. Also, the existence of vcpkg from Microsoft makes this task much less daunting, even if it doesn't include a WebKit package directly). A daunting task, but it might be less daunting than...
* The entire app could be rewritten into a display toolkit that cares about both cross-platform displays AND rich HTML displays.

**Comment by yetist** [comment]
In my opinion, xiphos needs to support both gtkhtml and webkit2gtk, and include gtkhtml as a third-party dependency fallback.

In an environment with webkit2gtk, html view uses webkit2gtk and html editor uses gtkhtml.
In an environment without webkit2gtk, html view uses gkhtml, and html editor uses gtkhtml.

Until one day, we completed the porting of html editor function to webkit2gtk, and webkit2gtk is effective on all platforms, we can delete the gtkhtml related code, and only keep the code of webkit2gtk.

**Comment by greg-hellings** [comment]
@yetist 

1) You're not going to get buy in from packaging folks to embed/vendor gtkhtml. It's just not going to happen. We're going to lose packages in basically every distro out there (maybe Arch excepted?)

2) webkit2gtk is not going to support Windows. The developers don't want to support it and they're not going to support it. WebKit's GTK port exists only to give Gnome a rich HTML widget for its personal vendetta to make Epiphany or whatever its successor might be exist.

So we either need to ditch WebKitGTK entirely, or ditch Windows entirely, if we want to remain up to date. In the world of weird things, I've found this library: https://github.com/zserge/webview which is a header-only library that embeds a native widget (WebKit2GTK on Linux, MSHTML/EdgeHTML/EdgeChromium on Windows, or WebKit on Mac OS) into the window you give it. There's lots of cross-platform options out there. It's now clear the WebKitGTK is not one of them.

**Comment by yetist** [comment]
> 1. You're not going to get buy in from packaging folks to embed/vendor gtkhtml. It's just not going to happen. We're going to lose packages in basically every distro out there (maybe Arch excepted?)

Archlinux also does not package gtkhtml, because it has stopped maintenance. I mean add the gtkhtml code to the xiphos code tree, such as 3rd-party/gtkhtml, and compile it together at build time.

**Comment by greg-hellings** [comment]
1) Arch still provides gtkhtml4-git, which can be installed from aur
2) That's exactly what I'm saying you're going to die on packaging from most any distro anywhere. There is a reason GtkHTML was removed (vulnerabilities, bugs, and no upstream maintenance) and those are the exact reasons that packagers refuse to permit bundled code.

So, if you want to maintain availability in mainline distros (I can speak for Fedora, and I've heard the same from Ubuntu, Debian, and SuSE), that option is a complete non-starter for running on those distributions.

**Comment by yetist** [comment]
For xiphos, we only need basic HTML display and editing functions, do not need too many functions, and do not need to depends on too large external libraries. Therefore, as long as xiphos can self-contain a small html library and realize html viewing and editing is enough.

The so-called web security issue does not apply to xiphos because its data source is trusted.

**Comment by greg-hellings** [comment]
Its data source is very much untrusted (anyone can create a Sword module and give it to anyone else to run; our modules are typically downloaded over FTP or possibly HTTP both of which are vulnerable to attacks) and none of that negates the fact that embedding a library is pretty much a terrible idea.

And I can't foresee Xiphos rolling its own HTML widget internally, for sure.

**Comment by alerque** [comment]
Pretty much have to echo what @greg-hellings is saying all the way down here. You're going to loose packaging on most distros if you start doing things like bundling outdated HTML engines from 3rd party sources, and that includes Arch Linux. It's just a _huge_ no-no.

Supporting _multiple_ outdated and unmaintained HTML renderers is crazy too. This is just an absurd amount of heft to lug around. We're already discussing the editor on the mailing list and dropping GTKHTML does seem like it might happen sooner rather than later.

From [this comment](https://github.com/crosswire/xiphos/issues/963#issuecomment-618171493), I'm not too keen to see 1 — but any of the three ways of achieving 2 seem viable.

**Comment by yetist** [comment]
OK, let's get back to the topic. Can the gtk 2 code be deleted?

**Comment by karlkleinpaste** [comment]
xiphos is completely in bed with gtk, there is no functional gtk3 for windows, and there are no immediate prospects for the complete replacement of either flavor of gtk in the near future.

ergo, not yet, and not any time soon.

**Comment by bgermann** [comment]
Has the Windows situation changed with GTK4?

**Comment by greg-hellings** [comment]
I have not yet found any evidence that it has changed. Keep in mind it mainly has 2 issues - both a GTK+ build for Windows, which has existed basically all along. And a build of WebKitGTK for Windows. Which they stopped doing a LONG time ago. Long ago enough that we still need to use GTK2 to tie into that version.

It doesn't seem that the WebKitGTK folks care about ever porting to Windows, frankly. If we want to ever modernize past GTK2 we would need a different HTML display solution.

**Comment by bgermann** [comment]
[vcpkg](https://github.com/microsoft/vcpkg/blob/master/ports/gtk/portfile.cmake) has a GTK4 port. Maybe using that would help. I guess WebKitGTK is a lost case. Where does the Windows build take its GTK2 version from?

**Comment by greg-hellings** [comment]
Lots of builds of GTK4 exist for Windows, so that's not the issue here. The problem is that WebKitGTK stopped supporting Windows in 2014 when they moved to WebKit2, and GtkHTML which was our fallback went out of maintenance in 2015. WebKitGTK for Windows now points people towards WebKitCairo, which is supposedly maintained and cross-platform. I haven't even looked into that to know if it's true.

Since we are not leveraging any advanced features of HTML and we are only using locally rendered content, it is unlikely that there is a huge in-app reason to move off the older code. However, it is beginning to fall out of Linux distro repositories as it is now riddled with less efficient code and a huge number of bugs and vulnerabilities. So all the pressure to move off the older code is external. But GTK does not seem interested in supporting an official way to cross-platform render HTML any longer.

If we want to move forward we would need to move to supporting multiple rendering engines based on our platform or to an entirely different solution as WebKitGTK is no longer viable.

**Comment by bgermann** [comment]
Would an editor component that does not have HTML rendering capabilites suffice? I guess there are plenty of GTK options in this case.

**Comment by greg-hellings** [comment]
You could use something like a Markdown editor which would then be rendered to HTML is one direction I've considered. But it doesn't help with the fact we don't even have a modern display widget in Windows. I'm not sure we ever even fully updated to WebKit2, did we? The last time I had looked at the editor APIs they had been mostly moved out to very different calls that needed to interact with the rendering subprocess, etc. But it's been a solid 2 years since I really went down that rabbit hole. So I don't know its exact status.

Another choice is to move to something that is better supported, like the Chromium Embedded Framework. That is the library designed to wrap around Blink/Chromium for use exactly as we use WebKit now. It has some demos of it being used in GTK apps, so it's a solid chance that we could get it to work well. But then we have *yet another* forklift of code to *yet another* rendering engine.

There are lots of options, each with benefits and drawbacks. Maybe we should start a discussion or Wiki page in the project to outline the different choices?

* Move away from HTML editing directly
* Move away from WebKitGTK on Windows builds
* Move away from WebKitGTK entirely

**Comment by TingPing** [comment]
I think your conclusions are mostly right, I just want to give slightly more context as a WebKit contributor.

WebKitGTK not supporting Windows is only because of lack of resources. I am confident patches providing such support would be accepted. But I do not expect any of you to take on that task.

While the WinCairo port exists and is maintained I don't think its as simple to use as WebKitGTK though and would take effort.

CEF also has no integration with GTK and you would have to do per-platform work yourself. The example code given in the CEF codebase only would work on X11 I believe. Actual integration like WebKitGTK would take a ton of work there.

Embedding web engines is a lot of work sadly.

**Comment by yetist** [comment]
I plan to maintain gtkhtml for gtk-3.0 compatibility and fix compatibility bugs, and do not plan to add new features.
Maybe this will help with porting xiphos to gtk-3.0 ?

**Comment by greg-hellings** [comment]
GTKHTML already is compatible with GTK-3. There isn't much chance of resurrecting it, though. Certainly not as a single person effort.

**Comment by TingPing** [comment]
GTKHTML has been umaintained for a decade.

**Comment by yetist** [comment]
> GTKHTML already is compatible with GTK-3. There isn't much chance of resurrecting it, though. Certainly not as a single person effort.

Fixed a lot of gtkhtml build warnings, maybe 4.12.0 will be released soon. https://github.com/yetist/gtkhtml

**Comment by bgermann** [comment]
I would like to bring up one suggestion that is currently not viable but given some more months seems like a good fit: servo. Editing support is [currently being worked on](https://github.com/servo/servo/pull/42633) and there is an experimental [servo-gtk](https://github.com/nacho/servo-gtk) binding.

**Comment by greg-hellings** [comment]
So the editing needs are now taken care of. The newly released GTK TextView based editor works. It doesn't necessarily pick up all the formatting that existed in the previous HTML syntax so it would require users to manually re-format some of their stuff. But it is present, it works, and it seems to be working well.

Getting rid of GTK 2.0 still would require finding a suitable HTML display widget that can be embedded to succeed WebKitGTK on Windows. There /might/ be GTK 3.0 support in the last released version of WebKitGTK for Windows but I wouldn't keep my fingers crossed for it. That would at least allow the unification of both systems onto GTK 3, but the wheel continues to turn beneath us and I'm sure pressures will begin to mount for moving to GTK 4.0 eventually.

**Comment by greg-hellings** [comment]
Follow along with my adventures to get GTK3 builds on Windows working properly.

[Here](https://github.com/crosswire/xiphos/pull/1360)

**Comment by LAfricain** [comment]
@greg-hellings did you try Claude.ai?

**Comment by greg-hellings** [comment]
@LAfricain To accomplish what? Right now I just need to test if the Windows build with GTK3 is working.

**Comment by LAfricain** [comment]
To help you in this task. Maybe it's already working, and that's would be great! But Claude.ai is really very very good for finding solutions...

**Comment by greg-hellings** [comment]
I'm familiar with Claude's abilities. I have at times been mandated to use it or face the wrath of the executive suite. But currently the question revolves entirely around whether GTK 3.0 on Windows from back in 2017 (which is the versions of the tools we are stuck using) works well. At various times we have experimented with GTK 3.0 on Windows in the past there were layout issues where it wasn't strictly a functional bug but it made our usage of the toolkit a UX bug.

Since we are unable to upgrade to newer versions of the libraries on Windows (WebKitGTK long ago dropped support for Windows, so Fedora 30 is our most up to date build environment) I wanted to experiment with whether our current build system can work well with GTK 3.

So far I've found that building against `mingw64-webkitgtk3` and `mingw64-gtk3` on Fedora 30 goes off without much of a hitch. It necessitates using the WebKit1 API instead of WebKit2 because of the underlying webkitgtk3 package. But we still have full support for the WK1, so that isn't a problem. Now it needs someone to do a functional test. I tried in Wine last night, but we're accessing some Windows API that Wine hasn't gotten cover for it in the installer. So I'll have to bust out an actual Windows install in order to give it a try.

There really isn't anything for Claude to do in this effort.

**Comment by LAfricain** [comment]
What do you want to test? the xiphos.exe or something else?

**Comment by greg-hellings** [comment]
I was able to verify basic functionality. The icons on the main app do not appear correctly and the editor wouldn't come up. But I haven't been able to put any time into troubleshooting either of those yet.

But I am not a native Windows user, so I'm not the best long term tester. And my Windows machine is a Windows on Arm from many years ago, so it is sometimes prone to its own, unique issues. So if anyone else wants to give feedback that would be appreciated.

**Comment by LAfricain** [comment]
> The icons on the main app do not appear correctly and the editor wouldn't come up

Confirm that!

Should the daily devotional with the calendar now also appear, or not? Right now, it's still using the old method where you have to enter the dates manually.

The bookmarks seem to use GTK3 because they're buttons instead of squares like before.

When you open certain external windows, Xiphos crashes—particularly the Settings window. Other than that, the interface looks much better!

In summary, there are three issues to solve:
 * Window opening
 * Icon display
 * Possibly the calendar in Daily Dev

**Comment by LAfricain** [comment]
Just wanted to let you know that I've managed to fix the main issues on Windows! There's just one minor issue left: the date isn't showing up in the Daily Dev. But it works with the calendar just like it does on Linux. I'm looking into it, but we could go ahead and merge the branches if you want. After that, there'll be some cleanup work to do, but it'll be fun! No more duplicates.
https://github.com/LAfricain/xiphos/actions/runs/31002534312

**Comment by greg-hellings** [comment]
The PR for migrating Windows over to GTK3 has now landed. It still uses the deprecated Webkit1 API, as the webkit-gtk for GTK3 back when Windows was still supported had not yet exposed Webkit2 APIs. However, it removes the need for GTK2 functionality.

I would say we hold off until the next actual release of Xiphos to see if there are a flurry of bug reports before we commit to actually deleting the GTK2 code entirely. However, it's now deprecated code that isn't really in use anywhere!

**Comment by LAfricain** [comment]
I think we're ready for a new version! I'm going to start working on cleaning up the code on a separate branch, and if we don't run into any major issues in a few months, we'll be ready.
OK?
@karlkleinpaste , can you set up a new release?

**Comment by greg-hellings** [comment]
@LAfricain To me #1362 should be a hard blocker for a new release. Edited modules don't display any text at all after an app restart for me (and not just in the Windows builds).

**Comment by LAfricain** [comment]
@greg-hellings See the answer  in #1362

**Comment by karlkleinpaste** [comment]
I intend to get a release out soon, but "soon" is fuzzy and imprecise. There are several things needed first. One is a considerable update to the manual to reflect many new features. That includes a completely new set of screenshots to reflect today's appearance under GNOME (as default environment, even though I myself don't use it). There's some code I want to robustify. I need to drop a request to all the past-known interface translators to ask them to do an update as well, as quite a few new strings are now present. Maybe some other smallish things. It'll be a few weeks of on-'n-off prep.

Unless there are other significant features or bugfixes that are also likely to land soon, I think we should consider the current state of things more or less frozen other than showstopper kinds of problems.

Oh: @LAfricain I am overdue to get you into the credits section of the main `About` widget. How would you like to appear there?

On a possibly related personal note, I am entering retirement soon. Last day of Real Work™ is Oct 21. It remains to be seen how much time I will spend with Xiphos after that, but I surely don't expect the amount of time to go down. I must say that having other people contributing -- not only filing bug reports, that is, having people contribute actual updates, fixes, and improvements -- is a huge motivating factor for me. I know there have been long dead times here, but when there's collaboration, this work is a much happier thing for me.

**Comment by LAfricain** [comment]
>One is a considerable update to the manual to reflect many new features.

I agree, I can help.

>Oh: @LAfricain I am overdue to get you into the credits section of the main About widget. How would you like to appear there?

Maybe you can add Cyrille Mattiuzzo?

>On a possibly related personal note, I am entering retirement soon. Last day of Real Work™ is Oct 21. It remains to be seen how much time I will spend with Xiphos after that, but I surely don't expect the amount of time to go down. I must say that having other people contributing -- not only filing bug reports, that is, having people contribute actual updates, fixes, and improvements -- is a huge motivating factor for me. I know there have been long dead times here, but when there's collaboration, this work is a much happier thing for me.

I agree that it's nicer to have some help! You'll have to thank Claude. :) I wish you a wonderful retirement.

**Comment by karlkleinpaste** [comment]
I don't suppose there's a more recent WebKit out there for Windows, is there?
A number of things (parallel header line stability, vertical multi-column scroll instead of horizontal) would be immediately solved.

**Comment by greg-hellings** [comment]
No. WebKitGTK officially dropped support for Windows builds a geological epoch ago and has officially said they have no intentions of restoring support unless some folks step up to support it.

There are a large number of people making very thin GTK wrappers around what is known as "WebView2", which is the Windows native HTML rendering widget at the base of the Edge browser. That seems to be the way that everyone except Chrome and Firefox are taking these days. Unlike the old days with IE, the change to Blink as the rendering engine (which is just the Chrome engine and a descendant of WebKit) for Edge has led Microsoft to greatly simplify embedding its engine into applications.

All of those are GTK 3 or GTK 4 based, and they're all pretty small. Which means we probably could wrap it directly ourselves with only the basic pieces of functionality that we need. I even found someone doing a wrapper that is API identical to WebKitGTK so it only takes one or two lines of ifdef-ing around it.

I would think that should be our road forward if the GTK3 code survives the Windows trial. It will reduce the amount of code we have to ship, reduce the deprecated behavior, and allow a greatly simplified app as we wouldn't need to mess around with many different versions of library cross multiplied with API versions. Now that we don't need to mess with integrating the external editor, all of that can be reduced in complexity.

**Comment by karlkleinpaste** [comment]
Just FYI.

I suspect your plan ↑above↑ is a fine choice, but nonetheless, having asked Grok, there appears to be something out there now.

<img width="776" height="972" alt="Image" src="https://github.com/user-attachments/assets/f5a1ad9b-2d82-4fa5-9f46-8aeaa195681d" />

**Comment by greg-hellings** [comment]
Yeah, the Cairo port has been around for a very long time. I haven't seen it successfully cross-compiled and when I've seen people try to embed it into a GTK widget in a general sense they usually give up.

Our goal isn't to make a generic widget so we might be successful. But the docs I've seen claim things like GCC is known to not work and you must compile with VisualStudio. That would be quite a shift for us, although not insurmountable. Everything else we depend on can support MSVC. And GHA gives us access to it on Windows build hosts.

The comments that it isn't production grade don't give me confidence.

**Comment by LAfricain** [comment]
This issue can be closed no?
